common/rpc-directory/multi-process-RPC-directory.mjs

```diff
@@ -0,0 +1,144 @@
+// Copyright 2021 HITCON Online Contributors
+// SPDX-License-Identifier: BSD-2-Clause
+
+// Boilerplate for getting require() in es module.
+import {createRequire} from 'module';
+const require = createRequire(import.meta.url);
+
+const grpc = require('@grpc/grpc-js');
+const protoLoader = require('@grpc/proto-loader');
+const path = require('path');
+
+import assert from 'assert';
+import Directory from './directory.mjs'
+import Handler from './handler.mjs'
+import { promisify } from 'util';
+import { fileURLToPath } from 'url';
+
+/**
+ * This class is handling all RPC calls in Multi Process version, which does IPC by gRPC.
+ */
+class MultiProcessRPCDirectory extends Directory {
+  /**
+   * Create the RPC Directory.
+   * @constructor
+   * @param {String} address - address of this RPC directory
+   */
+  constructor() {
+    super();
+    this.handlers = {};
+    this.remoteCallRPC = {};
+    this.remoteServiceAddress = {};
+
+    const packageDefinition = protoLoader.loadSync(path.dirname(fileURLToPath(import.meta.url)) + '/rpc.proto', {
+      keepCase: true,
+      longs: String,
+      enums: String,
+      defaults: true,
+      oneofs: true
+    });
+    this.rpcProto = grpc.loadPackageDefinition(packageDefinition).RPC;
+  }
+
+  /**
+   * Part of the constructor that needs to be async.
+   * This is created because constructor can't be async.
+   * Note that this is called right after the constructor.
+   */
+  async asyncConstruct() {
+  }
+
+  /**
+   * Create a gRPC server and publish it to Redis.
+   * Should only be called by standalone extension process.
+   * @param {String} port - port number
+   * @returns
+   */
+  createGrpcServer(name, port) {
+    let server = new grpc.Server();
+    server.addService(this.rpcProto.RPC.service, { callRPC: async (call, callback) => await this.responseGrpcCall.bind(this)(call, callback, this.callRPC) });
+    server.bindAsync('0.0.0.0:' + port, grpc.ServerCredentials.createInsecure(), () => {
+      server.start();
+      console.log("The gRPC server for " + name + " is running on port " + port + ".");
+    });
+  }
+
+  /**
+   * Method for responing gRPC, should only be called by gRPC server.
+   * @param {Object} call - The request content, contains the same signature as callRPC.
+   * @param {Function} callback - Callback function.
+   * @param {Object} handlers - The handler of this class.
+   */
+  async responseGrpcCall(call, callback, callRPC) {
+    callback(null, {
+      response: JSON.stringify(await callRPC.bind(this)(call.request.callerServiceName, call.request.serviceName, call.request.methodName, ...JSON.parse(call.request.args)))
```

Comment:
[Optional] We could consider having a callLocalRPC() and callRemoteRPC().
responseGrpcCall() would only call callLocalRPC(), callRPC() would call both. This way if there's any mismatch in server configuration, we'll not result in an infinite loop to ourself.